### PR TITLE
Performance improvements + benchmarks

### PR DESCRIPTION
**Pull request by jodzga**
Implemented optimizations:
1. Drain queue in SerialExecutor
2. Use LIFOBiPriorityQueue instead of FIFOPriorityQueue
3. Disable trampoline
4. Improve ParTaskImpl to decrease number of memory copies and number of allocations
5. Improve tracing be removing reference counting and use ArrayLists instead of HashMaps

Implemented benchmarks.

### CODE REVIEW

File: src/com/linkedin/parseq/internal/LIFOBiPriorityQueue.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * {@link TaskQueue} implementation that recognizes only two classes of priorities:
+ * <ul>
+ * <li>greater than {@link Priority#DEFAULT_PRIORITY}</li>
+ * <li>less or equal to {@link Priority#DEFAULT_PRIORITY}</li>
+ * </ul>
+ * Elements from first class are polled in FIFO order while elements from second class are polled
+ * in LIFO order. The rationale behind it is that elements with highest priorities will be mostly
+ * timeouts and we want to minimize variance in difference between time at which they were supposed
```
Comment (ssodhanilinkedin) [inline src/com/linkedin/parseq/internal/LIFOBiPriorityQueue.java]: Why don't we keep a separate queue for timeout tasks? Fundamentally timeout tasks should be handled at highest priority and not interfere with other tasks right?

File: src/com/linkedin/parseq/internal/SerialExecutor.java [hunk 2]
```diff
@@ -109,7 +111,38 @@ private void tryExecuteLoop() {
...
+          // executor waiting to be executed. Since _pendingCount can be changed in other threads
+          // in is possible to get _pendingCount.get() == 1 and _pendingCount.decrementAndGet() > 0
+          // to be true few lines below.
+          if (_pendingCount.get() == 1) {
+            _deactivationListener.deactivated();
+          }
+        } catch (Throwable t) {
+          _uncaughtExecutionHandler.uncaughtException(t);
+        } finally {
+          if (_pendingCount.decrementAndGet() == 0) {
```
Comment (ssodhanilinkedin) [inline src/com/linkedin/parseq/internal/SerialExecutor.java]: I would need more understanding of this piece of code.

File: contrib/parseq-benchmark/src/main/java/com/linkedin/parseq/EventsArrival.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.linkedin.parseq;
+
+import java.util.concurrent.TimeUnit;
+
+public interface EventsArrival {
```
Comment (jodzga) [inline contrib/parseq-benchmark/src/main/java/com/linkedin/parseq/EventsArrival.java]: done

File: contrib/parseq-benchmark/src/main/java/com/linkedin/parseq/HistogramSerializer.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.linkedin.parseq;
+
+import org.HdrHistogram.Histogram;
+
+public interface HistogramSerializer {
+
+  String serlialize(Histogram histogram);
```
Comment (jodzga) [inline contrib/parseq-benchmark/src/main/java/com/linkedin/parseq/HistogramSerializer.java]: Thanks for catching!

File: contrib/parseq-benchmark/src/main/java/com/linkedin/parseq/PoissonEventsArrival.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+public class PoissonEventsArrival implements EventsArrival {
+
+  private final Random _rand = new Random(System.nanoTime());
+  private final double _nanosToNextEventMean;
+
+  public PoissonEventsArrival(double events, TimeUnit perUnit) {
+    _nanosToNextEventMean = perUnit.toNanos(1) / events;
```
Comment (jodzga) [inline contrib/parseq-benchmark/src/main/java/com/linkedin/parseq/PoissonEventsArrival.java]: Nope, it has to be > 0, added arguments checking.

File: contrib/parseq-benchmark/src/main/java/org/HdrHistogram/Base64CompressedHistogramSerializer.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import com.linkedin.parseq.HistogramSerializer;
+
+
+public class Base64CompressedHistogramSerializer implements HistogramSerializer {
+
+  private ByteBuffer targetBuffer;
+
+  @Override
+  public synchronized String serlialize(Histogram histogram) {
+    int requiredBytes = histogram.getNeededByteBufferCapacity() + 16;
```
Comment (jodzga) [inline contrib/parseq-benchmark/src/main/java/org/HdrHistogram/Base64CompressedHistogramSerializer.java]: Thanks for suggestion.

File: src/com/linkedin/parseq/internal/LIFOBiPriorityQueue.java (hunk 1 above)
Comment (jodzga) [inline src/com/linkedin/parseq/internal/LIFOBiPriorityQueue.java]: We do. Timeouts will land on _highPriority queue.

File: src/com/linkedin/parseq/internal/SerialExecutor.java (hunk 2 above)
Comment (jodzga) [inline src/com/linkedin/parseq/internal/SerialExecutor.java]: Added javadoc.

File: src/com/linkedin/parseq/Engine.java [hunk 7]
```diff
@@ -50,6 +53,11 @@
   public static final String MAX_CONCURRENT_PLANS = "_MaxConcurrentPlans_";
   private static final int DEFUALT_MAX_CONCURRENT_PLANS = Integer.MAX_VALUE;
 
+  public static final String DRAIN_SERIAL_EXECUTOR_QUEUE = "_DrainSerialExecutorQueue_";
+  private static final boolean DEFUALT_DRAIN_SERIAL_EXECUTOR_QUEUE = true;
```
Comment (davidhoa) [inline src/com/linkedin/parseq/Engine.java]: DEFAULT is spelled wrong.

File: src/com/linkedin/parseq/ParSeqGlobalConfiguration.java [hunk 8]
```diff
@@ -58,4 +60,30 @@ public static boolean isCrossThreadStackTracesEnabled() {
   public static void setCrossThreadStackTracesEnabled(boolean enabled) {
     _crossThreadStackTracesEnabled = enabled;
   }
+
+  /**
+   * Returns true if trampoline is currently enabled.
+   *
+   * Trampoline prevents stack overflow in situation of extremely deep large ParSeq plans. Typically this problem
+   * does not occur and trampoline can be disabled allowing performance optimization.
```
Comment (davidhoa) [inline src/com/linkedin/parseq/ParSeqGlobalConfiguration.java]: I understand the intent of trampoline, but I'm having trouble seeing how it does that. It is added to an executor when this flag is on, how does that help? Is this documented somewhere?

File: src/com/linkedin/parseq/Engine.java (hunk 7 above)
Comment (jodzga) [inline src/com/linkedin/parseq/Engine.java]: Aaah, thanks for catching it!

File: src/com/linkedin/parseq/ParSeqGlobalConfiguration.java (hunk 8 above)
Comment (jodzga) [inline src/com/linkedin/parseq/ParSeqGlobalConfiguration.java]: This flag is used by Continuations class which I think is decently documented. Do you have questions about Continuations class?

File: contrib/parseq-benchmark/pom.xml [hunk 9]
```diff
@@ -15,7 +15,7 @@
         <dependency>
             <groupId>com.linkedin.parseq</groupId>
             <artifactId>parseq</artifactId>
-            <version>2.6.9</version>
+            <version>2.6.10</version>
         </dependency>
         <dependency>
             <groupId>com.linkedin.parseq</groupId>
```
Comment (ssodhanilinkedin) [inline contrib/parseq-benchmark/pom.xml]: this should 2.6.10 for parseq-batching.

Comment (ssodhanilinkedin) [inline contrib/parseq-benchmark/pom.xml]: also please verify in other pom.xml files once.

Comment (jodzga) [inline contrib/parseq-benchmark/pom.xml]: Thanks for catching this. Fixed and verified other dependencies.

File: src/com/linkedin/parseq/ParSeqGlobalConfiguration.java (hunk 8 above)
Comment (davidhoa) [inline src/com/linkedin/parseq/ParSeqGlobalConfiguration.java]: ah, yes, I missed the docs there. thanks.

**Pull request merged**